src/software/amazon/ion/impl/lite/IonIntLite.java

```diff
@@ -152,24 +163,28 @@ public void setValue(Number value)
         checkForLock();
         if (value == null)
         {
+            _actual_type = IntegerType.INT;
             doSetValue(0, true);
         }
         else
         {
             if (value instanceof BigInteger)
             {
+                _actual_type = IntegerType.BIG_INTEGER;
                 BigInteger big = (BigInteger) value;
                 doSetValue(big);
             }
             else if (value instanceof BigDecimal)
             {
+                _actual_type = IntegerType.BIG_INTEGER;
                 BigDecimal bd = (BigDecimal) value;
                 doSetValue(bd.toBigInteger());
             }
             else
             {
                 // XXX this is essentially a narrowing conversion
```

Comment:
IonJava provides these kinds of conversions all over the place, and I don't think any of them are really specified. What happens if the user provides a floating point number with a fractional component here? Or a floating point number too large for a long? Some rounding or truncation must occur.

The IonReader APIs also have this problem. The numeric *Value() APIs are documented to work for any numeric type, but what happens when you use IonReader.intValue() to read a double? Some undocumented (as far as I can see) conversion takes place, and data might be lost.